### DISCUSSION TITLE
What is the correct/best way to create a Cloudevents extension?

### DISCUSSION THREAD

**Discussion by e-backmark-ericsson** [cdfoundation/sig-events#60] [General]
Should we have our CDF SIG Events definitions as properties on the top level in the Cloudevent, or as data payload, or as a combination of the two?

The discussion started in this PR: https://github.com/cdfoundation/sig-events/pull/55#issuecomment-855975438, but should continue here.

**Comment by erkist** [comment]
The CloudEvents documentation [states](https://github.com/cloudevents/spec/blob/v1.0.1/primer.md#cloudevent-attribute-extensions)

> Extension attributes to the CloudEvent specification are meant to be additional metadata that needs to be included to help ensure proper routing and processing of the CloudEvent. Additional metadata for other purposes, that is related to the event itself and not needed in the transportation or processing of the CloudEvent, should instead be placed within the proper extensibility points of the event (data) itself.

So, my takeaway is that we will have very few attributes related to routing or processing, so most or all of our stuff should go into the data payload.

**Reply by m-linner-ericsson** [reply]
When it comes to routing and processing, does that mean that we should only have the [type](https://github.com/cloudevents/spec/blob/v1.0.1/spec.md#type) in the attributes?

**Comment by m-linner-ericsson** [comment]
Currently, our events look like this:
```json
{
  "data": {
    "pipeline": "data"
  },
  "datacontenttype": "application/json",
  "id": "8ba216bb-b443-4019-956e-f9b4d1700847",
  "pipelinerunerrors": "Hopfully no errors",
  "pipelinerunid": "pipe1",
  "pipelinerunname": "My Pipeline",
  "pipelinerunstatus": "Starting",
  "pipelinerunurl": "http://my-pipelinerunner",
  "source": "cde-cli",
  "specversion": "1.0",
  "time": "2021-07-16T14:30:33.1000937Z",
  "type": "cd.pipelinerun.started.v1"
}

```

but could also look like this:
```json
{
  "data": {
    "pipeline": "data",
    "pipelinerunerrors": "Hopfully no errors",
    "pipelinerunid": "pipe1",
    "pipelinerunname": "My Pipeline",
    "pipelinerunstatus": "Starting",
    "pipelinerunurl": "http://my-pipelinerunner",
  },
  "datacontenttype": "application/json",
  "id": "8ba216bb-b443-4019-956e-f9b4d1700847",
  "source": "cde-cli",
  "specversion": "1.0",
  "time": "2021-07-16T14:30:33.1000937Z",
  "type": "cd.pipelinerun.started.v1"
}
```

which I feel is more in line with the quote above from @erkist.
